R/orderly.R

```diff
@@ -0,0 +1,191 @@
+
+  expected <- unlist(lapply(dat$artefacts, "[[", "filenames"), FALSE, FALSE)
+
+  id <- outpack::outpack_id()
+  path <- file.path(root$path, "draft", name, id)
+  fs::dir_create(path)
+
+  fs::dir_create(file.path(path, dirname(inputs)))
+  fs::file_copy(file.path(src, inputs), path)
+
+  envir <- envir %||% .GlobalEnv
+  assert_is(envir, "environment")
+
+  withCallingHandlers({
+    outpack::outpack_packet_start(path, name, parameters = parameters,
+                                  id = id, root = root$outpack)
+    if (!is.null(parameters)) {
+      list2env(parameters, envir)
+    }
+    outpack::outpack_packet_file_mark(inputs)
+    outpack::outpack_packet_add_custom("orderly", custom_metadata,
+                                       custom_metadata_schema())
+    for (p in dat$packages) {
+      library(p, character.only = TRUE)
+    }
+    withr::with_dir(path, {
+      for (s in dat$sources) {
+        sys.source(s, envir = envir)
+      }
+    })
+    outpack::outpack_packet_run(dat$script, envir)
+    check_produced_files(path, expected, outpack::outpack_packet_file_list())
+    outpack::outpack_packet_end()
+    unlink(path, recursive = TRUE)
+  }, error = function(e) {
+    ## Eventually fail nicely here with mrc-3379
+    outpack::outpack_packet_cancel()
+  })
+
+  id
+}
+
+
+orderly_custom_metadata <- function(orderly_yml_dat) {
+  custom_artefacts <- lapply(orderly_yml_dat$artefacts, function(x)
+    list(description = scalar(x$description),
+         paths = x$filenames))
+  custom_packages <- orderly_yml_dat$packages %||% character()
+  custom_global <- list()
+
+  ## Not yet handled here: global, readme
+  custom_role <- data_frame(
+    path = c("orderly.yml", orderly_yml_dat$script,
+             orderly_yml_dat$sources,
+             orderly_yml_dat$resources),
+    role = c("orderly_yml", "script",
+             rep_along("source", orderly_yml_dat$sources),
+             rep_along("resource", orderly_yml_dat$resources)))
+
+  list(
+    artefacts = custom_artefacts,
+    packages = custom_packages,
+    global = custom_global,
+    role = custom_role,
+    displayname = scalar(orderly_yml_dat$displayname),
+    description = scalar(orderly_yml_dat$description),
+    custom = NULL)
+}
+
+
+check_produced_files <- function(path, expected, status) {
+  found <- file_exists(expected, workdir = path)
+  if (any(!found)) {
+    stop("Script did not produce expected artefacts: ",
+         paste(squote(expected[!found]), collapse = ", "))
+  }
+  extra <- setdiff(status$path[status$status == "unknown"], expected)
+  if (length(extra) > 0) {
+    ## TODO: once we get logging, this would go through that;
+    ## currently the format is different to orderly but that's fine.
+    ##
+    ## TODO: once mrc-3382 is fixed we can mark these to ignore in
+    ## outpack
+    message("Some extra files found: ",
+            paste(squote(extra), collapse = ", "))
+  }
+}
+
+
+check_parameters <- function(parameters, info) {
```

Comment:
I'm a bit confused by the splitting of responsibilities between `outpack` and `orderly2` - why can the former not do this kind of parameter validation?